fix(BID-356): Lot Page in Eigen not updating status

lordkiz opened the pull request:
<!--

➡️ Use a PR title in the form of  `type(PROJECT-XXXX): what changed`
➡️ Provide the Jira ticket in square brackets like [PROJECT-XXXX]

❗️ If this is a work in progress, remember to prefix it with [WIP] and/or open a draft PR instead of normal PR

-->

This PR resolves [BID-356]

### Description
Prefetching Auctions + associated `environment.retain(operation)` allows for stale records in auction related records. It is  essential that these records are kept fresh at all times. This PR removes prefetching for these sales.

<!-- Info, implementation, how to get there, before & after screenshots & videos, follow-up work, etc -->

### PR Checklist

- [x] I tested my changes on **iOS** / **Android**.
- [ ] I added screenshots or videos to illustrate my changes.
- [ ] I added Tests and Stories for my changes.
- [ ] I added an [app state migration].
- [ ] I hid my changes behind a [feature flag].

### To the reviewers 👀

- [ ] I would like **at least one** of the reviewers to run this PR on the simulator or device.

<details><summary>Changelog updates</summary>

### Changelog updates

<!-- 📝 Please fill out at least one of these sections. -->
<!-- ⓘ 'User-facing' changes will be published as release notes. -->
<!-- ⌫ Feel free to remove sections that don't apply. -->
<!-- • Write a markdown list or just a single paragraph, but stick to plain text. -->
<!-- 📖 eg. `Enable lotsByFollowedArtists - john` or `Fix phone input misalignment - mary`. -->
<!-- 🤷‍♂️ Replace this entire block with the hashtag `#nochangelog` to avoid updating the changelog. -->

#### Cross-platform user-facing changes

- Fix Lot Page in Eigen not updating status - kizito

#### iOS user-facing changes

-

#### Android user-facing changes

-

#### Dev changes

-

<!-- end_changelog_updates -->

</details>

Need help with something? Have a look at our [docs], or get in touch with us.

[app state migration]: ../blob/main/docs/adding_state_migrations.md
[feature flag]: ../blob/main/docs/developing_a_feature.md
[docs]: ../blob/main/docs/README.md


[BID-356]: https://artsyproduct.atlassian.net/browse/BID-356?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

olerichter00 commented on src/app/AppRegistry.tsx:
```diff
@@ -328,8 +328,8 @@ export const modules = defineModules({
   ArtworkSubmissionStatusFAQ: reactModule(ArtworkSubmissionStatusFAQ),
-  Auction: reactModule(SaleQueryRenderer, { fullBleed: true }, [SaleScreenQuery]),
-  Auctions: reactModule(SalesQueryRenderer, {}, [SalesScreenQuery]),
+  Auction: reactModule(SaleQueryRenderer, { fullBleed: true }),
+  Auctions: reactModule(SalesQueryRenderer),
```
Removing screen queries from the app registry prevents prefetching, but also results in warnings in the console because prefetching simply fails. If you want to disable prefetching for these screens you should look for places where the URL is prefetched.

lordkiz replied:
@olerichter00 I can see that the queries are optional. Passing empty array would also work but it is dirty. I do not understand what you mean by `where the URL is prefetched`. My understanding is that `environment.retain` would persist the data we get from these queries.

olerichter00 replied:
We use [prefetchUrl()](https://github.com/artsy/eigen/blob/6020f328d87260b33efe6da9c0bf773d23c721e7/src/app/utils/queryPrefetching.ts#L57) to prefetch queries for given URLs. This helper uses the routing algorithm to find the correct query to prefetch for a given URL. To prevent prefetching the lot/auction/sale screen, you'd probably want to remove [these](https://github.com/artsy/eigen/blob/6020f328d87260b33efe6da9c0bf773d23c721e7/src/app/Scenes/Home/Components/SalesRail.tsx#L79) two lines.

lordkiz replied:
@olerichter00 This won't fix the general issue as this would prefetch the individual sale on that rail only. Queries passed as part of Module props in AppRegistry are injected into the Component and at app start these queries are prefetched. We can safely remove the console.errors when `!module.Queries`, right?

olerichter00 replied:
[These](https://github.com/artsy/eigen/blob/6020f328d87260b33efe6da9c0bf773d23c721e7/src/app/Scenes/Home/Home.tsx#L78) are the only queries that are prefetched when the app starts. I don't know of any other mechanism that would prefetch queries at app start.

olerichter00 reviewed (changes requested):
I don't think this would fix the issue. Even though the queries are prefetched, entering the screen will trigger another request which should update the data in the store. 

You might want to add `force: true` to [SalesQueryRenderer](https://github.com/artsy/eigen/blob/5fbad7d8bb5420e5cf5730c32eb9e974bf89079c/src/app/Scenes/Sales/index.tsx#L126)'s cache config to make sure it's not using stale data.

lordkiz commented:
> I don't think this would fix the issue. Even though the queries are prefetched, entering the screen will trigger another request which should update the data in the store.
> 
> You might want to add `force: true` to [SalesQueryRenderer](https://github.com/artsy/eigen/blob/5fbad7d8bb5420e5cf5730c32eb9e974bf89079c/src/app/Scenes/Sales/index.tsx#L126)'s cache config to make sure it's not using stale data.

Thanks @olerichter00 . In another PR https://github.com/artsy/eigen/pull/6656/files#diff-3aabde61aed1fad2f6add85d35ce07c298ebe6ced62fbe4fa06df98c2a5585e7R92 , I have modified this to useLazyLoadQuery which has `force: true` as default. The persistence of stale records was greatly reduced.

olerichter00 commented:
Shall we close this one then?

lordkiz commented:
> Shall we close this one then?
@olerichter00 I won't say that. Individual Auction pages do not have stale records when I removed the prefetch query. As screens that would heavily depend on realtime data in coming PRs, prefetching on them would have undesired effects. Since the queries are optional, I removed them and I don't see any warnings. Maybe i am missing something.

olerichter00 commented:
> prefetching on them would have undesired effects. 

As I explained above, it wouldn't have any undesired effect because the query gets fired again as soon as the user enters the screen.

lordkiz commented:
> As I explained above, it wouldn't have any undesired effect because the query gets fired again as soon as the user enters the screen.

@olerichter00 Thanks for explaining. Before closing this, I want to be certain: as long as I have `force: true`, I do not need to worry about `environment.retain(data)` because the data will be updated no matter what? I am finding it difficult to ascertain this on the UI because it is not straightforward to observe the effects of relay caching.

olerichter00 commented:
Maybe this doc about [Relay Fetch Policies](https://relay.dev/docs/guided-tour/reusing-cached-data/fetch-policies/) might help to understand how it works. `store-and-network` is the fetch policy you want to use here in combination with the cache config `force: true`. 

The current caching setup is quite complicated (Relay store + additional network caching middleware). But these two configs will guarantee you fresh data.

lordkiz commented:
> Maybe this doc about [Relay Fetch Policies](https://relay.dev/docs/guided-tour/reusing-cached-data/fetch-policies/) might help to understand how it works. `store-and-network` is the fetch policy you want to use here in combination with the cache config `force: true`.
> 
> The current caching setup is quite complicated (Relay store + additional network caching middleware). But these two configs will guarantee you fresh data.
> 
> Even without query prefetching, you could run into a situation where the user sees the data from the cache when coming back to a screen.

Fair enough. Thanks @olerichter00

lordkiz commented:
Closing in favour of used `force: true` config in this PR:  https://github.com/artsy/eigen/pull/6656/files#diff-3aabde61aed1fad2f6add85d35ce07c298ebe6ced62fbe4fa06df98c2a5585e7R92

lordkiz commented:
@olerichter00 Reopening this PR. The issue is not improved. It improves with removal of prefetching for these screens.

olerichter00 commented:
@lordkiz ok, then let's deactivate prefetching for these two screens for now. Could you please let me know how to reproduce the bug and which screen hasn't been improved when changing the cache config? I'd like to investigate further to make sure this doesn't affect other screens.

lordkiz commented:
> @lordkiz ok, then let's deactivate prefetching for these two screens for now. Could you please let me know how to reproduce the bug and which screen hasn't been improved when changing the cache config? I'd like to investigate further to make sure this doesn't affect other screens.

Sure. For eg: Modifying an auction you have opened after the app is already running. Values remain stale even after reopening the auction page. You might need to reload the app to see the updated state.

olerichter00 commented:
> Sure. For eg: Modifying an auction you have opened after the app is already running. Values remain stale even after reopening the auction page. You might need to reload the app to see the updated state.

Thanks! Just one detail, how exactly are you reopening the auction page?

lordkiz commented:
> Thanks! Just one detail, how exactly are you reopening the auction page? I couldn't reproduce the bug like this, but maybe I just did it differently.

@olerichter00 By just going back to the auction page

olerichter00 commented:
> @olerichter00 By just going back to the auction page

But back from where? There is for example a difference if click the back button and then the lot again in the rail or if you use the bottom navigation. The first will re-render the screen (and reload the data) and the second won't. Maybe you could send a video or give clear steps on how to reproduce it?

lordkiz commented:
> But back from where? There is for example a difference if you click the back button and then the lot again in the rail or if you use the bottom navigation. The first will re-render the screen (and reload the data) and the second won't. Maybe you could send a video or give clear steps on how to reproduce it?

@olerichter00 
-From Homepage hit View All to land on auctions page.
 -Open an auction
-Open an artwork in the auction
-Go to auction-staging and edit the auction
-Go back all the way to Home.
-Click View all again
-Inspect the auctions: data is stale
- Also if you edit the auction while still on Homepage without ever going to the Auctions page at all. When you open the Auctions page and select an auction and inspect edited auction, you'd see the data is stale because it was fetched before you opened it.

olerichter00 commented:
I've adjusted the relay config for both screens (Sales/Auctions & Artwork) to the following in https://github.com/artsy/eigen/pull/6767 which fixed the issue for me. This will also make sure that even if a user revisits one of the screens, the data will be reloaded.

```
{
  fetchPolicy: "store-and-network",
  networkCacheConfig: { force: true },
}
```
